### PR TITLE
Implements the promised based getStats. Enables them for Safari and FF.

### PR DESCRIPTION
**Pull request by damencho**
Adds stats and audio levels for Safari. Enables the new getStats API for Firefox, that will get rid of the following warning:
'non-maplike pc.getStats access is deprecated, and will be removed in the near future! See http://w3c.github.io/webrtc-pc/#getstats-example for usage.'

### CODE REVIEW

File: modules/statistics/RTPStatsCollector.js [hunk 1]
```diff
@@ -940,3 +967,342 @@ StatsCollector.prototype.processAudioLevelReport = function() {
 };
 
 /* eslint-enable no-continue */
+
+/**
+ * New promised based getStats report processing.
+ * Tested with chrome, firefox and safari. Not switching it on for for chrome as
+ * frameRate stat is missing and calculating it using framesSent,
+ * gives values double the values seen in webrtc-internals.
```
Comment (gpolitis) [inline modules/statistics/RTPStatsCollector.js]: My hat is off to you for catching this 🥇

File: modules/statistics/RTPStatsCollector.js [hunk 2]
```diff
@@ -940,3 +967,342 @@ StatsCollector.prototype.processAudioLevelReport = function() {
...
+};
+
+/**
+ * Converts the value to a non-negative number.
+ * If the value is either invalid or negative then 0 will be returned.
+ * @param {*} v
+ * @return {number}
+ * @private
+ */
+StatsCollector.prototype.getNonNegativeValue = function(v) {
```
Comment (gpolitis) [inline modules/statistics/RTPStatsCollector.js]: We use this method to get the ssrc, the bytes, the packets and the lost packets (negative because of retransmissions, maybe?). These look like unsigned positive ints, so I'm a bit confused why we need this (I'm not debating whether we need it or not, I'm just trying to understand why we need it). So, are we flipping the sign in an effort to work around some webrtc bug? If so, can we add a small comment documenting this? If there's a bug report on the webrtc side of things, we can include that too.

File: modules/RTC/TraceablePeerConnection.js [hunk 3]
```diff
@@ -2314,10 +2335,15 @@ TraceablePeerConnection.prototype.getStats = function(callback, errback) {
...
-        // at least get native webrtc Safari available for use.
+    } else if (browser.isSafariWithWebrtc() || browser.isFirefox()) {
+        // uses the new Promise based getStats
+        this.peerconnection.getStats()
+            .then(callback)
+            .catch(errback || (() => {
+
+                // Making sure that getStats won't fail if error callback is
+                // not passed.
+            }));
```
Comment (virtuacoplenny) [inline modules/RTC/TraceablePeerConnection.js]: If this is promise based then maybe you should return this.peerconnection.getStats() so that in the future the callbacks can be removed. This would only be if nothing depends on TraceablePeerConnection#getStats not returning a promise.

File: modules/statistics/RTPStatsCollector.js [hunk 4]
```diff
@@ -722,6 +742,16 @@ StatsCollector.prototype.processStatsReport = function() {
         }
     }
 
+    this.eventEmitter.emit(
```
Comment (virtuacoplenny) [inline modules/statistics/RTPStatsCollector.js]: Why was this moved? It's close to where byteSentStats is done being calculated?

Honestly, I've always had a hard time understanding the 400 line block of multiple nested ifs that is processStatsReport. I'm glad you tried breaking this apart.

File: modules/statistics/RTPStatsCollector.js [hunk 5]
```diff
@@ -940,3 +967,342 @@ StatsCollector.prototype.processAudioLevelReport = function() {
 };
 
 /* eslint-enable no-continue */
+
+/**
+ * New promised based getStats report processing.
+ * Tested with chrome, firefox and safari. Not switching it on for for chrome as
```
Comment (virtuacoplenny) [inline modules/statistics/RTPStatsCollector.js]: *for for

File: modules/statistics/RTPStatsCollector.js [hunk 6]
```diff
@@ -242,6 +250,7 @@ export default function StatsCollector(
      * @private
      */
     this._getStatValue = this._defineGetStatValueMethod(keys);
+    this._getNewStatValue = this._defineNewGetStatValueMethod(keys);
```
Comment (virtuacoplenny) [inline modules/statistics/RTPStatsCollector.js]: Did you find it more clear to have separate instance variables for the get stats method instead of sharing one but conditionally selecting which _defineXXX got called?

File: modules/statistics/RTPStatsCollector.js [hunk 7]
```diff
@@ -2,6 +2,7 @@ import browser from '../browser';
 import { browsers } from 'js-utils';
```
Comment (virtuacoplenny) [inline modules/statistics/RTPStatsCollector.js]: Just had a thought. Tell me what you think: put the stats formatting into a separate module. That separates the collecting from the formatting. The separate module could then conditionally load the right formatter for rtpstatscollector to use (or the collector conditionally loads the right formatter).

File: modules/RTC/TraceablePeerConnection.js (hunk 3 above)
Comment (damencho) [reply modules/RTC/TraceablePeerConnection.js]: Currently, RTPStatsCollector implements both the promise and non-promise getStats. So we always call it passing callbacks. 
We can avoid passing callbacks, with very little code duplications here:
https://github.com/jitsi/lib-jitsi-meet/blob/e3adbca60887eacd964b8336fc477202e270117f/modules/statistics/RTPStatsCollector.js#L314
Do you want me to change it? But then TraceablePeerConnection.getStats in one case will depend on the callbacks and in the other case will just return Promise:
 * @returns {Promise|void}, is this ok?

File: modules/statistics/RTPStatsCollector.js (hunk 4 above)
Comment (damencho) [reply modules/statistics/RTPStatsCollector.js]: Emitting BYTE_SENT_STATS depends on local variable and I left it in that method, while extracting: this._processAndEmitReport();, as it is used in two places. 
I can also move the emit BYTE_SENT_STATS inside _processAndEmitReport, by passing byteSentStats to the _processAndEmitReport method. Does this sound ok?

File: modules/statistics/RTPStatsCollector.js (hunk 5 above)
Comment (damencho) [reply modules/statistics/RTPStatsCollector.js]: will fix it.

File: modules/RTC/TraceablePeerConnection.js [hunk 8]
```diff
@@ -531,6 +531,28 @@ TraceablePeerConnection.prototype.getTrackBySSRC = function(ssrc) {
...
 
+/**
+ * Tries to find SSRC number for given {@link JitsiTrack} id. It will search
+ * both local and remote tracks bound to this instance.
+ * @param {string} id
+ * @return {number|null}
+ */
+TraceablePeerConnection.prototype.getSsrcByTrackId = function(id) {
+
+    for (const localTrack of this.localTracks.values()) {
```
[the remark this answers is not in the document]
Comment (damencho) [reply modules/RTC/TraceablePeerConnection.js]: Will change that.

File: modules/RTC/TraceablePeerConnection.js (hunk 3 above)
Comment (virtuacoplenny) [reply modules/RTC/TraceablePeerConnection.js]: I guess we can wait if the promise isn't actually used yet. I was being too preemptive.

File: modules/statistics/RTPStatsCollector.js (hunk 6 above)
Comment (damencho) [reply modules/statistics/RTPStatsCollector.js]: Yep, makes sense. I was worried about spreading the if statements deciding which getStats implementation is used, but this thing can also be fixed. Will change that. Thanks.

File: modules/statistics/RTPStatsCollector.js (hunk 7 above)
Comment (damencho) [reply modules/statistics/RTPStatsCollector.js]: What do you mean by formatter? So currently what has been collected is been formatted in _processAndEmitReport which is common, no matter of the type of the stats...

File: modules/statistics/RTPStatsCollector.js (hunk 4 above)
Comment (virtuacoplenny) [reply modules/statistics/RTPStatsCollector.js]: No, it's okay where it is probably. I just couldn't see the intent of the move without some context.

File: modules/statistics/RTPStatsCollector.js (hunk 7 above)
Comment (virtuacoplenny) [reply modules/statistics/RTPStatsCollector.js]: I mean separating the collecting and the formatting logic. Conceptually they seem like two different concepts and the collecting logic doesn't necessarily need to know about formatting. That way the collector can use an arbitrary formatter, thereby hiding that there are two implementations for formatting.

Comment (virtuacoplenny) [reply modules/statistics/RTPStatsCollector.js]: But I now get the feeling, since you mentioned instance variable somewhere else, that the collector and formatting may not have the clean separation already that could allow the collector to take and use an arbitrary formatter.
